test/integration/buildrun_cleanup_test.go

```diff
@@ -0,0 +1,293 @@
+			_, err = tb.GetBRTillNotFound(buildRunObject.Name, time.Second*1, time.Minute)
+			Expect(apierrors.IsNotFound(err)).To(BeTrue())
+		})
+	})
+
+	Context("Multiple successful buildruns related to build with limit 1", func() {
+
+		BeforeEach(func() {
+			buildSample = []byte(test.MinimalBuildWithRetentionLimitOne)
+			buildRunSample = []byte(test.MinimalBuildRunRetention)
+		})
+
+		It("The older successful buildrun should not exist", func() {
+
+			Expect(tb.CreateBuild(buildObject)).To(BeNil())
+
+			buildObject, err = tb.GetBuildTillValidation(buildObject.Name)
+			Expect(err).To(BeNil())
+
+			// Create first buildrun
+			buildRunObject.Name = BUILDRUN + tb.Namespace + "-1"
+			Expect(tb.CreateBR(buildRunObject)).To(BeNil())
+			br1, err := tb.GetBRTillCompletion(buildRunObject.Name)
+			Expect(err).To(BeNil())
+			Expect(br1.Status.GetCondition(v1alpha1.Succeeded).Status).To(Equal(corev1.ConditionTrue))
+			// Create second buildrun
+			buildRunObject.Name = BUILDRUN + tb.Namespace + "-2"
+			Expect(tb.CreateBR(buildRunObject)).To(BeNil())
+			br2, err := tb.GetBRTillCompletion(buildRunObject.Name)
+			Expect(err).To(BeNil())
+			Expect(br2.Status.GetCondition(v1alpha1.Succeeded).Status).To(Equal(corev1.ConditionTrue))
+
+			_, err = tb.GetBR(BUILDRUN + tb.Namespace + "-1")
+			Expect(apierrors.IsNotFound(err)).To(BeTrue())
+		})
+
+	})
+
+	Context("Multiple buildruns with different build limits for failure and success", func() {
+
+		BeforeEach(func() {
+			cbsObjectFail, err = tb.Catalog.LoadCBSWithName(STRATEGY+tb.Namespace+"-fail", []byte(test.ClusterBuildStrategySingleStepNoPush))
+			Expect(err).To(BeNil())
+
+			// Load build strategy that can fail
+			err = tb.CreateClusterBuildStrategy(cbsObjectFail)
+			Expect(err).To(BeNil())
+			buildSample = []byte(test.MinimalBuildWithRetentionLimitDiff)
+			buildRunSample = []byte(test.MinimalBuildRunRetention)
+
+		})
+
+		AfterEach(func() {
+			err := tb.DeleteClusterBuildStrategy(cbsObjectFail.Name)
+			Expect(err).To(BeNil())
+		})
+		// Case with failedLimit 1 and succeededLimit 2.
+		// It ensures that only relevant buildruns are affected by retention parameters
+		It("Should not find the old failed buildrun if the limit has been triggered", func() {
+
+			// Create build
+			Expect(tb.CreateBuild(buildObject)).To(BeNil())
+			buildObject, err = tb.GetBuildTillValidation(buildObject.Name)
+			Expect(err).To(BeNil())
+
+			// Create 2 successful buildruns
+			buildRunObject.Name = BUILDRUN + tb.Namespace + "-success-1"
+			err = tb.CreateBR(buildRunObject)
+
+			buildRunObject.Name = BUILDRUN + tb.Namespace + "-success-2"
+			err = tb.CreateBR(buildRunObject)
+
+			// Load and create the failing buildobject with the relevant cbs
+			buildObject, err = tb.Catalog.LoadBuildWithNameAndStrategy(BUILD+tb.Namespace+"-fail", STRATEGY+tb.Namespace+"-fail", buildSample)
+			Expect(err).To(BeNil())
+			buildObject.Spec.Source.ContextDir = nil
+			Expect(tb.CreateBuild(buildObject)).To(BeNil())
```

Comment:
if we move into a noop strategy that can be parameterized to tweak beahviour via BuildRuns, I think you dont need to create another `Build`, one `Build` should be enough.